### DISCUSSION TITLE
Linking chgres_cube failed on macOS Monterey

### DISCUSSION THREAD

**Discussion by ywangwof** [ufs-community/ufs-srweather-app#717] [Q&A] [answered]
I am trying to build **ufs-srweather-app** using script _devbuild.sh_ on macOS Monterey using GNU compilers and got the following message

> [ 59%] Linking Fortran executable chgres_cube
> ld: library not found for -lmpi_cxx
> collect2: error: ld returned 1 exit status
> make[5]: *** [sorc/chgres_cube.fd/chgres_cube] Error 1
> make[4]: *** [sorc/chgres_cube.fd/CMakeFiles/chgres_cube.dir/all] Error 2
> make[3]: *** [all] Error 2
> make[2]: *** [sorc/UFS_UTILS/src/UFS_UTILS-stamp/UFS_UTILS-build] Error 2
> make[1]: *** [sorc/CMakeFiles/UFS_UTILS.dir/all] Error 2
> make: *** [all] Error 2
> 

**hpc-stack** has been installed with **openmpi 4.1.2** successfully and both **ufs_model** and **upp** are built without a problem.

Anyone has the same problem before and a solution? Thanks.

**Comment by natalie-perlin** [comment]
@ywangwof - thank you for sharing your experience, and the details you provided - all are needed! I'll be looking into it. Few more questions on your setup: are you using Mac with _M1_ or _Intel_ chip? What branch of **ufs-srweather-app** are you using (develop, release/public-v2.1.0)?

**Reply by ywangwof** [reply]
My Mac is with intel chip and the **ufs-srweather-app** branch is develop by this commit.

> * baf4df1d - (2023-04-07) [develop] Point source pre-combine (#708)

**Comment by ywangwof** [comment]
I found the problem is that library `-lmpi_cxx` is not in the linking paths for program **chgres_cube** (it is installed in `$HPC_INSTALL_DIR/gnu-12.2.0/openmpi/4.1.2/lib`). There are two solutions
1. Copy/Link the library from `$HPC_INSTALL_DIR/gnu-12.2.0/openmpi/4.1.2/lib/libmpi_cxx.dylib` to `$BREW/Cellar/gcc/12.2.0/bin/../lib/gcc/current/gcc/x86_64-apple-darwin21/12/../../../`.
2. Put `$HPC_INSTALL_DIR/gnu-12.2.0/openmpi/4.1.2/lib` to the linking paths of program **chgres_cube**.

Solution 1 is not generalized for all users. About solution 2, I do not know how to modify the _CMakeList.txt_ file for it to work. Any suggestion? Thanks.

**Reply by natalie-perlin** [reply, marked as answer]
@ywangwof - I do attempt to reproduce your error. 
As to the library not in the path: after you load the modulefile,
**module load build_macos_gnu**
there is a message asking you to export an LDFLAGS variable:
`export LDFLAGS+=" -L$MPI_ROOT/lib " `
(It is likely that the message in the modulefiles needs to be corrected to display exactly the one shown above)

This is where the library lmpi_cxx is expected to be found. For the MacOS, you are likely to have the following files in your $MPI_ROOT/lib directory: libmpi_cxx.la, libmpi_cxx.dylib

**Reply by ywangwof** [reply]
Thanks @natalie-perlin. Your answer explains the problem. I was using _devbuild.sh_ to build the programs, which load the module automatically and it does not give users a chance to read the output message from module "**build_macos_gnu**". 

I am wondering whether it is possible to set the environment variable **LDFLAGS** in the script _devbuild.sh_ automatically after it loads the module file on macOS?

**Comment by natalie-perlin** [comment]
@ywangwof - thanks for letting me know it became more clear!
The documentation for [building the SRW v2.1 on MacOS](https://ufs-srweather-app.readthedocs.io/en/release-public-v2.1.0/BuildSRW.html#additional-details-for-building-on-macos-or-generic-linux) indeed suggests using [CMake approach](https://ufs-srweather-app.readthedocs.io/en/release-public-v2.1.0/BuildSRW.html#cmake-approach) vs. devbuils.sh . 
And yes, adding the LDFLAGS by default sounds like a better approach, so I'd consider implementing these changes when possible. Thank you for your involvement!
